### PR TITLE
Adds structure-RerF as a `methodToUse`

### PR DESCRIPTION
**Pull request by MrAE**
I have added structured-RerF for 2d gray-scale images as `methodToUse=2`.

I have also added a test to make sure the features selected comprise a contiguous patch in the image.

### CODE REVIEW

File: packedForest/src/forestTypes/binnedTree/processingNodeBin.h [hunk 1]
```diff
@@ -80,28 +79,106 @@ namespace fp{
 
 
 				inline void calcMtryForNode(std::vector<weightedFeature>& featuresToTry){
-					//add this to fpInfo so user can set if needed.
-					int methodToUse = 1;
 					featuresToTry.resize(fpSingleton::getSingleton().returnMtry());
+					int methodToUse = fpSingleton::getSingleton().returnMethodToUse();
+					assert(methodToUse == 1 || methodToUse == 2);
```
Comment (falkben) [inline packedForest/src/forestTypes/binnedTree/processingNodeBin.h:84]: `raise` instead?

File: packedForest/src/forestTypes/binnedTree/processingNodeBin.h [hunk 2]
```diff
@@ -80,28 +79,106 @@ namespace fp{
...
+						case 1:{
+							randMatTernary(featuresToTry);
+							break;
+							}
+						case 2:{
+							std::vector<int> patchPositions = paramsRandMatImagePatch();
+							randMatImagePatch(featuresToTry, patchPositions);
+							break;
 							}
+						default:{
```
Comment (falkben) [inline packedForest/src/forestTypes/binnedTree/processingNodeBin.h]: any reason for this?  my thinking is better to not do things by default (see above about raising an error) better to be explicit

File: packedForest/src/forestTypes/binnedTree/processingNodeBin.h [hunk 3]
```diff
@@ -80,28 +79,106 @@ namespace fp{
...
+					// Preset parameters
+					const int& imageHeight = fpSingleton::getSingleton().returnImageHeight();
+					const int& imageWidth = fpSingleton::getSingleton().returnImageWidth();
+
+					const int& patchHeightMax = fpSingleton::getSingleton().returnPatchHeightMax();
+					const int& patchHeightMin = fpSingleton::getSingleton().returnPatchHeightMin();
+					const int& patchWidthMax  = fpSingleton::getSingleton().returnPatchWidthMax();
+					const int& patchWidthMin  = fpSingleton::getSingleton().returnPatchWidthMin();
+
+					std::vector<int > heightWidthTop(3 * fpSingleton::getSingleton().returnMtry());
```
Comment (falkben) [inline packedForest/src/forestTypes/binnedTree/processingNodeBin.h]: why 3 times mtry?  can you add a comment?

File: packedForest/src/forestTypes/binnedTree/processingNodeBin.h [hunk 4]
```diff
@@ -80,28 +79,106 @@ namespace fp{
...
+						// compute the difference between the image dimensions and the current random patch dimensions for sampling
+						deltaH = imageHeight - heightWidthTop[k] + 1;
+						deltaW = imageWidth - heightWidthTop[k + 1] + 1;
+
+						// Sample the top left pixel from the available pixels (due to buffering).
+						topLeftSeed = randNum->gen(deltaH * deltaW);
+
+						// Convert the top-left-seed value to it's appropriate index in the full image.
+						heightWidthTop[k + 2] = (topLeftSeed % deltaW) + (imageWidth * floor(topLeftSeed / deltaW));
+						assert(heightWidthTop[k] < (imageHeight * imageWidth));
```
Comment (falkben) [inline packedForest/src/forestTypes/binnedTree/processingNodeBin.h]: look into this (possibly k+2 instead)

File: packedForest/src/forestTypes/binnedTree/processingNodeBin.h [hunk 5]
```diff
@@ -80,28 +79,106 @@ namespace fp{
...
+
+				inline void randMatImagePatch(std::vector<weightedFeature>& featuresToTry, std::vector<int> patchPositions){
+					assert(int (patchPositions.size()) == (3*fpSingleton::getSingleton().returnMtry()));
+
+					// Preset parameters
+					//const int& imageHeight = fpSingleton::getSingleton().returnImageHeight();
+					const int& imageWidth = fpSingleton::getSingleton().returnImageWidth();
+
+					int pixelIndex = -1;
+					for (int k = 0; k < (3*fpSingleton::getSingleton().returnMtry()); k += 3){
```
Comment (falkben) [inline packedForest/src/forestTypes/binnedTree/processingNodeBin.h]: can we rename j/i to row/col ?

File: packedForest/src/forestTypes/binnedTree/processingNodeBin.h [hunk 6]
```diff
@@ -80,28 +79,106 @@ namespace fp{
...
+						for (int j = 0; j < patchPositions[k]; j++) {
+							for (int i = 0; i < patchPositions[k + 1]; i++) {
+								pixelIndex = patchPositions[k + 2] + i + (imageWidth * j);
```
Comment (falkben) [inline packedForest/src/forestTypes/binnedTree/processingNodeBin.h]: //indexing all the pixels in patch

File: packedForest/src/fpSingleton/fpInfo.h [hunk 7]
```diff
@@ -71,6 +82,13 @@ namespace fp {
 				float_epsilon=std::numeric_limits<float>::epsilon()*10;
 				seed=-1;
 				numTreeBins=-1;
+				methodToUse = 1; // Should this default to 1?
```
Comment (falkben) [inline packedForest/src/fpSingleton/fpInfo.h:85]: add these parameter defaults to the constructor

line ~258

File: packedForest/test/fpTests/binnedForest/processingNodeBinTest.h [hunk 8]
```diff
@@ -98,4 +98,66 @@ TEST(testProcessingNodeBin, testMtryMult)
...
+	fpSingleton::getSingleton().setParameter("imageWidth", 28);
+	fpSingleton::getSingleton().setParameter("patchHeightMax", 28);
+	fpSingleton::getSingleton().setParameter("patchHeightMin", 1);
+	fpSingleton::getSingleton().setParameter("patchWidthMax", 28);
+	fpSingleton::getSingleton().setParameter("patchWidthMin", 1);
+	fpSingleton::getSingleton().loadData();
+	fpSingleton::getSingleton().setDataDependentParameters();
+	fpSingleton::getSingleton().checkDataDependentParameters();
+
+	int numClasses = fpSingleton::getSingleton().returnNumClasses();
```
Comment (falkben) [inline packedForest/test/fpTests/binnedForest/processingNodeBinTest.h]: unused

File: packedForest/test/fpTests/fpSingletonTest.h [hunk 9]
```diff
@@ -0,0 +1,54 @@
+#include "../../src/fpSingleton/fpSingleton.h"
+
+using namespace fp;
+
+TEST(fpSingleton, checkDataDependantParameters_errorPatchHeightMax)
```
Comment (falkben) [inline packedForest/test/fpTests/fpSingletonTest.h:5]: please add test for `paramsRandMatImagePatch`

given width = height = 28, assert top = 0
given width = height = 27, assert top = 0,1,28,29
given min/max of different sizes, sampled height/widths are within range

**Review by falkben — CHANGES_REQUESTED** [review]
awesome!  minor fixes

File: packedForest/src/forestTypes/binnedTree/processingNodeBin.h (hunk 1 above)
Comment (MrAE) [reply packedForest/src/forestTypes/binnedTree/processingNodeBin.h:84]: If I leave the assert there -- mostly for dev use -- and add a throw statement at the time `methodToUse` is set then we won't have to check it at every node and that'd be better, right?

Comment (falkben) [reply packedForest/src/forestTypes/binnedTree/processingNodeBin.h:84]: agreed

File: packedForest/src/fpSingleton/fpInfo.h (hunk 7 above)
Comment (MrAE) [reply packedForest/src/fpSingleton/fpInfo.h:85]: I now get these warnings:

```
Synapse:packedForest JLP$ make
g++-8 -std=c++11 -fopenmp -Wall -O3 -DNDEBUG -ffast-math -Iinc -I/usr/local/include  -c -o obj/fp.o src/fp.cpp
In file included from src/fpSingleton/fpData.h:7,
                 from src/fpSingleton/fpSingleton.h:4,
                 from src/packedForest.h:22,
                 from src/fp.cpp:1:
src/fpSingleton/fpInfo.h: In constructor 'fp::fpInfo::fpInfo()':
src/fpSingleton/fpInfo.h:61:9: warning: 'fp::fpInfo::useRowMajor' will be initialized after [-Wreorder]
    bool useRowMajor;
         ^~~~~~~~~~~
src/fpSingleton/fpInfo.h:31:8: warning:   'int fp::fpInfo::methodToUse' [-Wreorder]
    int methodToUse;
        ^~~~~~~~~~~
src/fpSingleton/fpInfo.h:255:4: warning:   when initialized here [-Wreorder]
    fpInfo(): numTreesInForest(100),
    ^~~~~~
src/fpSingleton/fpInfo.h:37:8: warning: 'fp::fpInfo::patchHeightMax' will be initialized after [-Wreorder]
    int patchHeightMax;
        ^~~~~~~~~~~~~~
src/fpSingleton/fpInfo.h:36:8: warning:   'int fp::fpInfo::patchHeightMin' [-Wreorder]
    int patchHeightMin;
        ^~~~~~~~~~~~~~
src/fpSingleton/fpInfo.h:255:4: warning:   when initialized here [-Wreorder]
    fpInfo(): numTreesInForest(100),
    ^~~~~~
src/fpSingleton/fpInfo.h:39:8: warning: 'fp::fpInfo::patchWidthMax' will be initialized after [-Wreorder]
    int patchWidthMax;
        ^~~~~~~~~~~~~
src/fpSingleton/fpInfo.h:38:8: warning:   'int fp::fpInfo::patchWidthMin' [-Wreorder]
    int patchWidthMin;
        ^~~~~~~~~~~~~
src/fpSingleton/fpInfo.h:255:4: warning:   when initialized here [-Wreorder]
    fpInfo(): numTreesInForest(100),
    ^~~~~~
g++-8 -std=c++11 -o bin/fp obj/fp.o -fopenmp -lm -O3 
```

Comment (MrAE) [reply packedForest/src/fpSingleton/fpInfo.h:85]: I fixed it.  I ordered them incorrectly

**Review by falkben — APPROVED** [review]
nice! i like the `ASSERT_THROW`s

**Pull request merged**